### DISCUSSION TITLE
Some questions about the network layer of parking env (SAC+HER)

### DISCUSSION THREAD

**Discussion by xunpla123** [eleurent/highway-env#161] [General]
Hi, i tried to run the `parking env` and got the success rate of over 95%. But I still have some questions about the model structure.
The model as follows : 
```
model = HER('MlpPolicy', env, SAC, n_sampled_goal=4,
            goal_selection_strategy='future',
            verbose=1, buffer_size=int(1e6),
            learning_rate=1e-3, gamma=0.9, batch_size=256,
            policy_kwargs=dict(layers=[256, 256, 256]))          
```
I'm not really understanding about `policy_kwargs`. I think it is the parameter to setup model or create policy and target TF objects, but i am not sure. This model does not only have one network makes me confuse. 
May i ask you about the neural network architecture of this env ? If you have time to answer, I hope to figure out how many layers are there and  how many neurons in each layer.

Looking forward to your reply, thanks.

**Comment by eleurent** [comment]
Hi @xunpla123, 

> I'm not really understanding about policy_kwargs. I think it is the parameter to setup model or create policy and target TF objects, but i am not sure

Yes, exactly: `policy_kwargs` is a dictionary used to configure the policy network, whose type is `MlpPolicy` (1st argument).

> This model does not only have one network makes me confuse.
> May i ask you about the neural network architecture of this env ? If you have time to answer, I hope to figure out how many layers are there and how many neurons in each layer.

The model is of type `MlpPolicy`, which stands for Multi-Layer Perceptron. Its architecture is defined in the `layers` parameter of `policy_kwargs`: three layers of 256 neurons each.

**Comment by xunpla123** [comment]
Thanks for your reply ! May i ask other questions ?
I think the input of the SAC network structure are `env.observation_space` and `env.action_space`. Is it correct ? If it is, i want to know their dimensions.

**Comment by eleurent** [comment]
`env.observation_space` defines the inputs of the policy, while `env.action_space` defines its outputs.

* `env.observation_space` is typically a continuous `Box`, and you can get its dimension by calling `env.observation_space.shape`
* `env.action_space` can be a `Box`, but also a `Discrete` space, in which case you can get its size by calling  `env.action_space.n`

**Comment by xunpla123** [comment]
Have you tried other policy model ? (`CnnPolicy`, `LnMlpPolicy`, ...) 
Would they get different results from `MlpPolicy` ?

**Comment by eleurent** [comment]
No, for this particular task I already had satisfying results with `Mlp` so I did not try more sophisticated models (plus the observation is quite small and unstructured).

For the intersection environment however, I compared MLP with CNN, as well as an attention-based model, see [paper](https://arxiv.org/abs/1911.12250) and [results](https://eleurent.github.io/social-attention/) (with a runnable colab notebook).

**Comment by xunpla123** [comment]
Hi @eleurent, may i ask you a new question about the layers ?

I want to know why you decided to use the 3 layers network. I try to use the more complex networks ( 6 layers or 9 layers ... ) , then they may not converge .  ( The mean reward and success rate will increase at first, but will decrease in the middle of the training. )

**Comment by eleurent** [comment]
No particular reason, fully-connected networks with 2 or 3 layers are a common baseline in RL papers.
I did not perform any kind of hyperparameter optimization though, there may be many other architectures that perform better.
